CORE-17779 add metric for a HTTP RPC Client request

LWogan opened the pull request:
Add a metric to record some information about how long it takes for an RPCClient to receive a response, as well as a distribution of response message size

ben-millar committed:
CORE-16181 Adding RPCClient along with associated factories and unit tests

ben-millar committed:
CORE-16181 Adding RPCClient along with associated factories and unit tests

ben-millar committed:
CORE-16181 Adding RPCClient along with associated factories and unit tests

ben-millar committed:
CORE-16181 Adding RPCClient along with associated factories and unit tests

ben-millar committed:
CORE-16181 Reverting StateManager override patch

ben-millar committed:
CORE-16181 Reverting StateManager override patch

ben-millar committed:
CORE-16181 Reverting StateManager override patch

ben-millar committed:
CORE-16181 Reverting StateManager override patch

ben-millar committed:
CORE-16181 Adding RPC routing to FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Adding RPC routing to FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Adding RPC routing to FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Adding RPC routing to FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Defining HTTP exceptions in the correct place

ben-millar committed:
CORE-16181 Defining HTTP exceptions in the correct place

ben-millar committed:
CORE-16181 Defining HTTP exceptions in the correct place

ben-millar committed:
CORE-16181 Defining HTTP exceptions in the correct place

ben-millar committed:
CORE-16181 RPCClient now retrieves endpoint from message

ben-millar committed:
CORE-16181 RPCClient now retrieves endpoint from message

ben-millar committed:
CORE-16181 RPCClient now retrieves endpoint from message

ben-millar committed:
CORE-16181 RPCClient now retrieves endpoint from message

ben-millar committed:
CORE-16181 Moved worker RPC paths to constants file

ben-millar committed:
CORE-16181 Moved worker RPC paths to constants file

ben-millar committed:
CORE-16181 Moved worker RPC paths to constants file

ben-millar committed:
CORE-16181 Moved worker RPC paths to constants file

ben-millar committed:
CORE-16181 Implemented HTTP call retry functionality with unit tests

ben-millar committed:
CORE-16181 Implemented HTTP call retry functionality with unit tests

ben-millar committed:
CORE-16181 Implemented HTTP call retry functionality with unit tests

ben-millar committed:
CORE-16181 Implemented HTTP call retry functionality with unit tests

ben-millar committed:
CORE-16181 Added helper function for retrieving endpoint from message

ben-millar committed:
CORE-16181 Added helper function for retrieving endpoint from message

ben-millar committed:
CORE-16181 Added helper function for retrieving endpoint from message

ben-millar committed:
CORE-16181 Added helper function for retrieving endpoint from message

ben-millar committed:
CORE-16181 Removed unnecessary test, made assertThrows exceptions more precise

ben-millar committed:
CORE-16181 Removed unnecessary test, made assertThrows exceptions more precise

ben-millar committed:
CORE-16181 Removed unnecessary test, made assertThrows exceptions more precise

ben-millar committed:
CORE-16181 Removed unnecessary test, made assertThrows exceptions more precise

ben-millar committed:
CORE-16181 Fixing detekt errors

ben-millar committed:
CORE-16181 Fixing detekt errors

ben-millar committed:
CORE-16181 Fixing detekt errors

ben-millar committed:
CORE-16181 Fixing detekt errors

mbrkic-r3 committed:
CORE-16203 Replace State and Event pattern with Multi-Source Event Mediator in FlowWorker (#4832)

Using Multi-Source Event Mediator instead of State and Event Subscription in FlowWorker.

Polling consumers and committing offsets performed in mediator thread in a synchronous way (without using coroutines)

mbrkic-r3 committed:
CORE-17429 State metadata support in message processor (#4828)

Adds support for state metadata to the state and event processor.

Components using the event mediator to process events and states may wish to set metadata alongside the state that is written back to the underlying state storage. This change adds to the StateAndEventProcessor API to facilitate this, as well as migrating existing state and event processors to use this.

---------

Co-authored-by: James Higgs <james.higgs@r3.com>

ben-millar committed:
CORE-16181 Adding RPCClient to FlowEventMediatorFactory clientFactories list

ben-millar committed:
CORE-16181 Adding RPCClient to FlowEventMediatorFactory clientFactories list

ben-millar committed:
CORE-16181 Adding RPCClient to FlowEventMediatorFactory clientFactories list

ben-millar committed:
CORE-16181 Adding RPCClient to FlowEventMediatorFactory clientFactories list

ben-millar committed:
CORE-16181 Moving HTTPRetryConfig/Executor from messaging to messaging-impl

ben-millar committed:
CORE-16181 Moving HTTPRetryConfig/Executor from messaging to messaging-impl

ben-millar committed:
CORE-16181 Moving HTTPRetryConfig/Executor from messaging to messaging-impl

ben-millar committed:
CORE-16181 Moving HTTPRetryConfig/Executor from messaging to messaging-impl

ben-millar committed:
CORE-16181 Removing RPC endpoints which aren't yet implemented

ben-millar committed:
CORE-16181 Removing RPC endpoints which aren't yet implemented

ben-millar committed:
CORE-16181 Removing RPC endpoints which aren't yet implemented

ben-millar committed:
CORE-16181 Removing RPC endpoints which aren't yet implemented

ben-millar committed:
CORE-16181 Reverting commits which removed changes to the router in FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Reverting commits which removed changes to the router in FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Reverting commits which removed changes to the router in FlowEventMediatorFactoryImpl

ben-millar committed:
CORE-16181 Reverting commits which removed changes to the router in FlowEventMediatorFactoryImpl

mbrkic-r3 committed:
CORE-16203 Removed coroutine usage from Multi-Source Event Mediator.

mbrkic-r3 committed:
CORE-16203 Set linger.ms to 0 for stateAndEvent producer in kafka-messaging-defaults.conf

mbrkic-r3 committed:
CORE-17487 5.1 Performance integration - parallel processing (#4873)

Add support for parallel flows executing within the flow pipeline.

This includes:
- Temporary multi-thread implementation of the state and event pattern.
- Using the SameThreadExecutor for flow fibers, which causes the fiber to run on the thread that started the fiber.
- Using a Kryo Pool for Kryo instances, as kryo instances are not thread-safe themselves.

Add TaskManager and a TaskManagerFactory which is a wrapper around a ScheduledExecutorService used for gathering metrics on executing tasks. The built-in Micrometer metrics for executor services are used, along with some additional metrics in the TaskManager that are more specific to our use-cases.
---------

Co-authored-by: Conal Smith <conal.smith@r3.com>
Co-authored-by: Dan Newton <dan.newton@r3.com>
Co-authored-by: Dries Samyn <dries.samyn@r3.com>

mbrkic-r3 committed:
CORE-17661 5.1 Performance integration - Multi-Source Event Mediator integration with StateManager (#4874)

Integration of Multi-Source Event Mediator with StateManager.

Commit:
Merge pull request #4872 from corda/mbrkic-r3/feature/51-perf-integration

CORE-15419 5.1 Performace Integration

mbrkic-r3 committed:
CORE-16203 5.1 Performance integration - FlowMapper using Multi-Source Event Mediator (#4875)

Using Multi-Source Event Mediator instead of State and Event Subscription in FlowMapperWorker.

---------

Co-authored-by: James Higgs <james.higgs@r3.com>

Omar-awad committed:
Merge pull request #4876 from corda/mbrkic-r3/feature/51-perf-integration-2

CORE-15419 5.1 Performance Integration #2

mbrkic-r3 committed:
CORE-17562 Metrics for Multi-Source Event Mediator

mbrkic-r3 committed:
CORE-17562 Fixed unit test

Commit:
Merge branch 'release/os/5.1' into feature/51-perf-integration

Commit:
Update API version to fix build (#4882)

ben-millar committed:
CORE-16181 Rerouting ledger events to RPC client

ben-millar committed:
CORE-16181 Rerouting ledger events to RPC client

ben-millar committed:
CORE-16181 Rerouting ledger events to RPC client

ben-millar committed:
CORE-16181 Rerouting ledger events to RPC client

ben-millar committed:
CORE-16181 Removing coroutines from RPCClient and HTTPRetryExecutor

ben-millar committed:
CORE-16181 Removing coroutines from RPCClient and HTTPRetryExecutor

ben-millar committed:
CORE-16181 Removing coroutines from RPCClient and HTTPRetryExecutor

ben-millar committed:
CORE-16181 Removing coroutines from RPCClient and HTTPRetryExecutor

ben-millar committed:
CORE-16181 Removing ser/des from RPCClient as its handled at subscription level

ben-millar committed:
CORE-16181 Removing ser/des from RPCClient as its handled at subscription level

ben-millar committed:
CORE-16181 Removing ser/des from RPCClient as its handled at subscription level

ben-millar committed:
CORE-16181 Removing ser/des from RPCClient as its handled at subscription level

ben-millar committed:
CORE-16181 Adding /api/5.1/ prefix to worker paths

ben-millar committed:
CORE-16181 Adding /api/5.1/ prefix to worker paths

ben-millar committed:
CORE-16181 Adding /api/5.1/ prefix to worker paths

ben-millar committed:
CORE-16181 Adding /api/5.1/ prefix to worker paths

ben-millar committed:
CORE-16181 Removing coroutines include

ben-millar committed:
CORE-16181 Removing coroutines include

ben-millar committed:
CORE-16181 Removing coroutines include

ben-millar committed:
CORE-16181 Removing coroutines include

ben-millar committed:
CORE-16181 Fixing unit test data format

ben-millar committed:
CORE-16181 Fixing unit test data format

ben-millar committed:
CORE-16181 Fixing unit test data format

ben-millar committed:
CORE-16181 Fixing unit test data format

ben-millar committed:
CORE-16181 Creating a HttpClientFactory so the same client can be reused across RPCClients

ben-millar committed:
CORE-16181 Creating a HttpClientFactory so the same client can be reused across RPCClients

ben-millar committed:
CORE-16181 Creating a HttpClientFactory so the same client can be reused across RPCClients

ben-millar committed:
CORE-16181 Creating a HttpClientFactory so the same client can be reused across RPCClients

Omar-awad committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17562/mediator-metrics-2

ben-millar committed:
CORE-16181 Pulling API version from platformInfoProvider

ben-millar committed:
CORE-16181 Pulling API version from platformInfoProvider

ben-millar committed:
CORE-16181 Pulling API version from platformInfoProvider

ben-millar committed:
CORE-16181 Pulling API version from platformInfoProvider

ben-millar committed:
CORE-16181 Updating RPCClient request from PUT to POST

ben-millar committed:
CORE-16181 Updating RPCClient request from PUT to POST

ben-millar committed:
CORE-16181 Updating RPCClient request from PUT to POST

ben-millar committed:
CORE-16181 Updating RPCClient request from PUT to POST

Omar-awad committed:
Merge pull request #4879 from corda/mbrkic-r3/CORE-17562/mediator-metrics-2

CORE-17562 Metrics for Multi-Source Event Mediator

ben-millar committed:
CORE-16181 Adding CordaDeserializer to RPCClient

ben-millar committed:
CORE-16181 Adding CordaDeserializer to RPCClient

ben-millar committed:
CORE-16181 Adding CordaDeserializer to RPCClient

ben-millar committed:
CORE-16181 Adding CordaDeserializer to RPCClient

ben-millar committed:
CORE-16181 Adding timeout to HTTPClient

ben-millar committed:
CORE-16181 Adding timeout to HTTPClient

ben-millar committed:
CORE-16181 Adding timeout to HTTPClient

ben-millar committed:
CORE-16181 Adding timeout to HTTPClient

thiagoviana committed:
CORE-16242 - Synchronous RPC Pattern API Implementation for Crypto Worker (#4820)

HTTP RPC server integration into the Crypto processor. Allows Crypto processor to receive HTTP external event requests sent from the flow engine. Client integration will come later.

ben-millar committed:
CORE-16181 Adding PlatformInfoProvider to integ tests

ben-millar committed:
CORE-16181 Adding PlatformInfoProvider to integ tests

ben-millar committed:
CORE-16181 Adding PlatformInfoProvider to integ tests

ben-millar committed:
CORE-16181 Adding PlatformInfoProvider to integ tests

ben-millar committed:
CORE-16181 Adding new Crypto endpoint to FlowEvent message router

ben-millar committed:
CORE-16181 Adding new Crypto endpoint to FlowEvent message router

ben-millar committed:
CORE-16181 Adding new Crypto endpoint to FlowEvent message router

ben-millar committed:
CORE-16181 Adding new Crypto endpoint to FlowEvent message router

ben-millar committed:
CORE-16181 Fixing detekt issue

ben-millar committed:
CORE-16181 Fixing detekt issue

ben-millar committed:
CORE-16181 Fixing detekt issue

ben-millar committed:
CORE-16181 Fixing detekt issue

ben-millar committed:
CORE-16181 Removing serializer from RPCClient, changing des type to Any()

ben-millar committed:
CORE-16181 Removing serializer from RPCClient, changing des type to Any()

ben-millar committed:
CORE-16181 Removing serializer from RPCClient, changing des type to Any()

ben-millar committed:
CORE-16181 Removing serializer from RPCClient, changing des type to Any()

mbrkic-r3 committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4894)

* CORE-16181 Fixed issue with setting wrong state version in event mediator

mbrkic-r3 committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4894)

* CORE-16181 Fixed issue with setting wrong state version in event mediator

mbrkic-r3 committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4894)

* CORE-16181 Fixed issue with setting wrong state version in event mediator

mbrkic-r3 committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4894)

* CORE-16181 Fixed issue with setting wrong state version in event mediator

davidcurrie committed:
CORE-17822 Apply prefix to topic overrides (#4893)

LWogan committed:
Merge branch 'bm/CORE-16181/rpc-client' of https://github.com/corda/corda-runtime-os into bm/CORE-16181/rpc-client

JamesHR3 committed:
CORE-17627: Add metadata to flow mapper states for mapper status (#4886)

Adds the state status key to flow mapper state metadata and populates it with the current state.

The metadata key is used by the mapper cleanup logic to ensure that mapper states are removed.

ben-millar committed:
CORE-16181 Changing HTTPRetryExecutor logging from info to trace

ben-millar committed:
CORE-16181 Changing HTTPRetryExecutor logging from info to trace

LWogan committed:
CORE-17779 add metric for a HTTP RPC Client request

LWogan committed:
Merge remote-tracking branch 'origin/bm/CORE-16181/rpc-client' into CORE-17779/RPC-Client-Metrics

# Conflicts:
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/HTTPRetryExecutor.kt

LWogan committed:
CORE-17779 kdoc

ben-millar committed:
CORE-16181 Addressing PR comments

JamesHR3 committed:
CORE-17388: Add session timeout metadata to store alongside the checkpoint (#4889)

Adds session timeout to the metadata stored alongside the checkpoint.

Session timeout is implemented using the state storage lookup mechanism, which requires the metadata to be populated with the session timeout key. The earliest timestamp of session expiry for currently open sessions is the value assigned to this property.

LWogan committed:
CORE-17779 add separate metric for response size

LWogan committed:
Merge remote-tracking branch 'origin/bm/CORE-16181/rpc-client' into CORE-17779/RPC-Client-Metrics

# Conflicts:
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/factory/RPCClientFactory.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/HTTPRetryConfig.kt
#	libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/utils/HTTPRetryExecutorTest.kt

LWogan committed:
CORE-17779 fix merge conflicts

LWogan committed:
CORE-17779 more merge conflicts

LWogan committed:
CORE-17779 more merge conflicts

LWogan committed:
CORE-17779 cleanup code

Omar-awad reviewed (dismissed):
LTGM

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt:
```diff
@@ -73,8 +79,43 @@ class RPCClient(
+        request: HttpRequest,
+        response: HttpResponse<ByteArray>? = null
+    ) {
+        val endTime = Instant.now().toEpochMilli()
```
Think we might have been using `System.nanoTime()` isntead e.g. https://github.com/jujoramos/corda-runtime-os/blob/e86a998e6ca8799066ad10a930b6e440414d99a7/components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImpl.kt#L71

JamesHR3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt:
```diff
@@ -73,8 +79,43 @@ class RPCClient(
+    ) {
+        val endTime = Instant.now().toEpochMilli()
+        CordaMetrics.Metric.Messaging.HTTPRPCResponseTime.builder()
+            .withTag(CordaMetrics.Tag.OperationStatus, operationStatus)
```
Is it worth having both the operation status and the response code as tags? In theory that would increase the cardinality of our metrics if we ever decided to treat a 2xx code as an error (not sure why we'd do that). Feels like we're better off doing one or the other?

LWogan replied:
the operation status of failed indicates an exception was thrown so its possible no response was received

JamesHR3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt:
```diff
@@ -73,8 +79,43 @@ class RPCClient(
+            .record(endTime - startTime, TimeUnit.MILLISECONDS)
+
+        if (response != null) {
+            CordaMetrics.Metric.Messaging.HTTPRPCResponseSize.builder()
```
I'd be tempted to leave this one out entirely. If it is kept, I think we only record this in successful cases anyway so it might be worth taking off things like success/failure and the response code.

LWogan replied:
ive left it in and taken off the tags suggested for removal

ben-millar committed:
CORE-16181 Further addressing PR comments

JamesHR3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt:
```diff
@@ -73,8 +79,43 @@ class RPCClient(
+        CordaMetrics.Metric.Messaging.HTTPRPCResponseTime.builder()
+            .withTag(CordaMetrics.Tag.OperationStatus, operationStatus)
+            .withTag(CordaMetrics.Tag.HttpMethod, request.method())
+            .withTag(CordaMetrics.Tag.HttpRequestUri, request.uri().toString())
```
Are any of these URIs templated? We'll get a cardinality spike if we record a separate time series for each possible templated value in the URI.

LWogan replied:
for now the paths are very simple with no unique identifiers. In the format of ${server-url}/${WorkerPath}.

JamesHR3 commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -692,6 +703,16 @@ object CordaMetrics {
          */
         HttpMethod("http.method"),
 
+        /**
+         * The URI that a HTTP request was sent to
+         */
+        HttpRequestUri("http.request.uri"),
+
+        /**
+         * Response code received for a HTTP response
+         */
+        HttpResponseCode("http.response.code"),
+
```
Do we not already have similar tag names for use in the REST processor that we could reuse here?

LWogan replied:
i dont see any

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt:
```diff
@@ -73,8 +79,43 @@ class RPCClient(
+        val endTime = Instant.now().toEpochMilli()
+        CordaMetrics.Metric.Messaging.HTTPRPCResponseTime.builder()
+            .withTag(CordaMetrics.Tag.OperationStatus, operationStatus)
+            .withTag(CordaMetrics.Tag.HttpMethod, request.method())
```
Wonder will HttpMethod be useful here or does it just increase metric cardinality, maybe the value for HttpRequestUri could combine `request.method() + request.uri().toString()`. Wouldn't hold up the PR for it

LWogan committed:
Merge remote-tracking branch 'origin/bm/CORE-16181/rpc-client' into CORE-17779/RPC-Client-Metrics

# Conflicts:
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt

ben-millar committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4885)

This PR adds a new implementation of MessagingClient, RPCClient, handling synchronous calls to the Corda workers via RPC/HTTP with configurable retries.

Includes also a small bug-fix for the task manager.

---------

Co-authored-by: Miljenko Brkic <97448832+mbrkic-r3@users.noreply.github.com>

JamesHR3 committed:
CORE-17626: Add integration test for flow mapper cleanup (#4907)

Adds an integration test for mapper cleanup.

Also addresses a small bug where the key for the cleanup event was not serialized correctly.

LWogan committed:
Merge remote-tracking branch 'origin/feature/51-perf-integration' into CORE-17779/RPC-Client-Metrics

# Conflicts:
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/HTTPRetryExecutor.kt

LWogan committed:
CORE-16184 code review comments

LWogan committed:
CORE-16184 merge conflict

JamesHR3 committed:
CORE-17882: Add a route for flow events (#4915)

Adds a route for flow events to the flow worker event mediator.

LWogan committed:
Merge branch 'feature/51-perf-integration' into CORE-17779/RPC-Client-Metrics

Commit:
Merge branch 'release/os/5.1' into feature/51-perf-integration

LWogan committed:
Merge branch 'feature/51-perf-integration' into CORE-17779/RPC-Client-Metrics

LWogan committed:
Merge remote-tracking branch 'origin/feature/51-perf-integration' into CORE-17779/RPC-Client-Metrics

LWogan committed:
Merge remote-tracking branch 'origin/CORE-17779/RPC-Client-Metrics' into CORE-17779/RPC-Client-Metrics

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into CORE-17779/RPC-Client-Metrics

# Conflicts:
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/messaging/FlowEventMediatorFactoryImplTest.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/RPCClient.kt

LWogan committed:
CORE-16184 merge conflict

Omar-awad reviewed (approved):
LGTM
